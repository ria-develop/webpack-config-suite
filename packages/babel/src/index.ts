export * from './JsxExtension';
export * from './HotModuleReplacementConfig';
export * from './BabelReactConfig';
export * from './BabelConfig';
