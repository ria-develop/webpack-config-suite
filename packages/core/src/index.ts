import {Configuration} from 'webpack';
import {BaseConfig} from './BaseConfig';

export * from './BaseConfig';
export * from './Entry';
export * from './EntryObject';
export * from './ModuleRules';
export * from './Plugins';
export * from './Resolve';
export * from './ResolveAlias';
export * from './ResolveExtensions';
export * from './ResolvePlugins';
export * from './Utils';
export * from './WebpackSetup';
export * from './impl/CacheThreadRule';
export * from './impl/JsExtension';

export type {LoaderOptions as ExtractCssLoaderOptions} from '../types/mini-css-extract-loader';
export type {Options as CacheLoaderOptions} from '../types/cache-loader';
export type {Options as CssLoaderOptions} from '../types/css-loader';
export type {Options as LessLoaderOptions} from '../types/less-loader';
export type {Options as PostCssLoaderOptions} from '../types/postcss-loader';
export type {Options as StyleLoaderOptions} from '../types/style-loader';
export type {PluginOptions as MiniCssExtractPluginOptions} from '../types/mini-css-extract-plugin';
export type {Schema as BabelLoaderOptions} from '../types/babel-loader';

export type WebpackCliOptions = {
  mode?: Configuration['mode'];
  liveReload?: boolean;
  hot?: boolean;
  hotOnly?: boolean;
  [key: string]: any;
};

export type What = typeof BaseConfig | ConfigHandler | (typeof BaseConfig | ConfigHandler)[];
export type ConfigHandler = (config: Configuration, env: unknown, argv: WebpackCliOptions) => What[] | void;
