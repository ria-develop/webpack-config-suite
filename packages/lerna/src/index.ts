export * from './LernaOutput';
export * from './LernaTypeScriptPlugins';
export * from './LernaTypeScriptReactConfig';
export * from './LernaTypeScriptResolveModules';
export * from './LernaTypeScriptResolvePlugins';
export * from './Utils';
