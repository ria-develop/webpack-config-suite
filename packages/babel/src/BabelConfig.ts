import {
  BabelLoaderOptions as BabelLoaderOptionsFromSchema,
  BaseConfig,
  CacheThreadRule,
  getLoader,
  JsExtension
} from '@webpack-config-suite/core';
import {RuleSetUseItem} from 'webpack';
import {PluginItem} from '@babel/core';
import {Options as BabelPresetEnvOptions} from '@babel/preset-env';

import {TransformOptions} from 'babel__core';

export type BabelLoaderOptions = TransformOptions & BabelLoaderOptionsFromSchema;

export class BabelConfig extends CacheThreadRule {
  static BABEL_PRESET_ENV: '@babel/preset-env' = '@babel/preset-env';

  static BABEL_PRESET_REACT: '@babel/preset-react' = '@babel/preset-react';

  static BABEL_PLUGIN_HOT_LOADER: 'react-hot-loader/babel' = 'react-hot-loader/babel';

  public get composed(): typeof BaseConfig[] {
    return [JsExtension];
  }

  //--------------------------------------------------------------------------------------------------------------------
  // Preset Env
  //--------------------------------------------------------------------------------------------------------------------
  protected get useReact(): boolean {
    return false;
  }

  protected get babelPresetEnvOptions(): BabelPresetEnvOptions | null {
    return null;
  }

  protected get babelPresetEnv(): (string | BabelPresetEnvOptions | null)[] {
    return [BabelConfig.BABEL_PRESET_ENV, this.babelPresetEnvOptions].filter(Boolean);
  }

  //--------------------------------------------------------------------------------------------------------------------
  // Preset React
  //--------------------------------------------------------------------------------------------------------------------

  protected get babelPresetReactOptions(): unknown {
    return null;
  }

  protected get babelPresetReact(): unknown[] {
    return [BabelConfig.BABEL_PRESET_REACT, this.babelPresetReactOptions].filter(Boolean);
  }

  //--------------------------------------------------------------------------------------------------------------------
  // General
  //--------------------------------------------------------------------------------------------------------------------

  protected get test(): RegExp | string {
    return /\.js$/;
  }

  protected get babelPresets(): PluginItem[] | null {
    return [this.babelPresetEnv, this.useReact && this.babelPresetReact].filter(Boolean);
  }

  protected get babelPlugins(): PluginItem[] | null {
    return this.isHot && [BabelConfig.BABEL_PLUGIN_HOT_LOADER];
  }

  protected get babelLoaderOptions(): BabelLoaderOptions {
    return {presets: this.babelPresets, ...(this.babelPlugins && {plugins: this.babelPlugins})};
  }

  protected get babelLoader(): RuleSetUseItem {
    return getLoader('babel-loader', this.babelLoaderOptions);
  }

  protected get use(): RuleSetUseItem[] {
    return super.use.concat([this.babelLoader]);
  }
}
