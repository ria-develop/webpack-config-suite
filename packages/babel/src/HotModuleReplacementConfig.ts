import {AliasAwareConfig, WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';

export class HotModuleReplacementConfig extends AliasAwareConfig {
  protected get alias(): Configuration['resolve']['alias'] {
    return this.isHot && {'react-dom': '@hot-loader/react-dom'};
  }

  constructor(config: Configuration, env: unknown, argv: WebpackCliOptions) {
    super(config, env, argv);
    if (this.isHot) {
      // why @see https://github.com/webpack/webpack-dev-server/issues/2758
      config.target = 'web';
    }
  }
}
