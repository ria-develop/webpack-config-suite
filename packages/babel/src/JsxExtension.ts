import {ResolveExtensions} from '@webpack-config-suite/core';

export class JsxExtension extends ResolveExtensions {
  protected get extensions(): string[] {
    return ['.jsx'];
  }
}
