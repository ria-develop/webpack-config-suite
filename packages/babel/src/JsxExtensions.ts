import {JsExtension} from '@webpack-config-suite/core';

export class JsxExtensions extends JsExtension {
  protected get extensions(): string[] {
    return super['extensions'].concat('.jsx');
  }
}
export default JsxExtensions;
