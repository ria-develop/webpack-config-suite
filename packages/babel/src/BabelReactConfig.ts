import {BaseConfig} from '@webpack-config-suite/core';
import {BabelConfig} from './BabelConfig';
import {HotModuleReplacementConfig, JsxExtension} from './';

export class BabelReactConfig extends BabelConfig {
  public get composed(): typeof BaseConfig[] {
    return super.composed.concat([JsxExtension, this.isHot && HotModuleReplacementConfig]).filter(Boolean);
  }

  protected get test(): RegExp | string {
    return /\.jsx?$/;
  }

  protected get useReact(): boolean {
    return true;
  }
}
