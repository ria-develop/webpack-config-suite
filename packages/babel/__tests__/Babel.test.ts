import {WebpackCliOptions} from '@webpack-config-suite/core';
import {Configuration} from 'webpack';
import {BabelConfig, BabelReactConfig, HotModuleReplacementConfig} from '../src';

describe.each([
  ['development', undefined, {}],
  ['development', undefined, {hot: true}]
])(
  'given mode: "%s", env: "%j", argv: "%j"',
  (mode: Configuration['mode'], env: unknown, argv: WebpackCliOptions): void => {
    it('given BabelConfig should match snapshot', () => {
      const config: Configuration = {mode};
      class MyBabelConfig extends BabelConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }
      new MyBabelConfig(config, env, argv).composed.forEach((Config) => new Config(config, env, argv));
      expect(config).toMatchSnapshot();
    });
    it('given BabelReactConfig should match snapshot', () => {
      const config: Configuration = {mode};
      class MyBabelReactConfig extends BabelReactConfig {
        protected get numWorkers(): number {
          return 1;
        }
      }
      new MyBabelReactConfig(config, env, argv).composed.forEach((Config) => new Config(config, env, argv));
      expect(config).toMatchSnapshot();
    });
    it('given HotModuleReplacementConfig should match snapshot', () => {
      const config: Configuration = {mode};

      new HotModuleReplacementConfig(config, env, argv);
      expect(config).toMatchSnapshot();
    });
  }
);
