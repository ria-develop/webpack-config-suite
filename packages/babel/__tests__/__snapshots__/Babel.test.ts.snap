// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`given mode: "development", env: "undefined", argv: "{"hot":true}" given BabelConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "plugins": Array [
                "react-hot-loader/babel",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"hot":true}" given BabelReactConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "plugins": Array [
                "react-hot-loader/babel",
              ],
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "alias": Object {
      "react-dom": "@hot-loader/react-dom",
    },
    "extensions": Array [
      ".jsx",
    ],
  },
  "target": "web",
}
`;

exports[`given mode: "development", env: "undefined", argv: "{"hot":true}" given HotModuleReplacementConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "resolve": Object {
    "alias": Object {
      "react-dom": "@hot-loader/react-dom",
    },
  },
  "target": "web",
}
`;

exports[`given mode: "development", env: "undefined", argv: "{}" given BabelConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.js\\$/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {},
}
`;

exports[`given mode: "development", env: "undefined", argv: "{}" given BabelReactConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "module": Object {
    "rules": Array [
      Object {
        "exclude": /node_modules/,
        "test": /\\\\\\.jsx\\?\\$/,
        "use": Array [
          Object {
            "loader": "cache-loader",
          },
          Object {
            "loader": "thread-loader",
            "options": Object {
              "poolTimeout": Infinity,
              "workers": 1,
            },
          },
          Object {
            "loader": "babel-loader",
            "options": Object {
              "presets": Array [
                Array [
                  "@babel/preset-env",
                ],
                Array [
                  "@babel/preset-react",
                ],
              ],
            },
          },
        ],
      },
    ],
  },
  "resolve": Object {
    "extensions": Array [
      ".jsx",
    ],
  },
}
`;

exports[`given mode: "development", env: "undefined", argv: "{}" given HotModuleReplacementConfig should match snapshot 1`] = `
Object {
  "mode": "development",
  "resolve": Object {},
}
`;
